### PR TITLE
[dagster-dbt] Support @dbt_assets with select in a dbt-core>=1.8 project with unit tests

### PR DESCRIPTION
**Pull request by maximearmstrong**
## Summary & Motivation

This PR fixes issue #22854.

### Problem

An exception is raised when initializing `@dbt_assets` with a selector method "tag" passed to either `select` or `exclude`, only when the dbt project includes a `unit_test.yml` file.

The exception is caused because `node.tags == None` when running [this line of code](https://github.com/dbt-labs/dbt-core/blob/v1.8.4/core/dbt/graph/selector_methods.py#L266-L271) in `dbt.graph.selector_methods`.

Adding a tag in `unit_test.yml` does not fixed the problem. Eg.

```
unit_tests:
  - name: test_unit_test
    config:
      tags:
        - test
    model: my model
```

Taking a closer look, `node.__class__  == 'dagster_dbt.utils.select_unique_ids_from_manifest.<locals>._DictShim'>` when running the selector method code [here](https://github.com/dbt-labs/dbt-core/blob/v1.8.4/core/dbt/graph/selector_methods.py#L266-L271). The `_DictShim` class is defined in `dagster-dbt` [here](https://github.com/dagster-io/dagster/blob/1.7.14/python_modules/libraries/dagster-dbt/dagster_dbt/utils.py#L233).

The class is expected to be one of the SelectorTarget classes, see [here](https://github.com/dbt-labs/dbt-core/blob/v1.8.4/core/dbt/graph/selector_methods.py#L105-L107).

For unit test, the class is expected to be UnitTestDefinition, see [here](https://github.com/dbt-labs/dbt-core/blob/1a2d367e7f09cdc8e8af77f0646ebbd07742a451/core/dbt/contracts/graph/manifest.py#L827). The root problem is that `UnitTestDefinition` has a property names `tags`, see [here](https://github.com/dbt-labs/dbt-core/blob/1a2d367e7f09cdc8e8af77f0646ebbd07742a451/core/dbt/contracts/graph/nodes.py#L966-L969), but `_DictShim` will return `None` because of how its `__get_attr__()` is implemented.

For the other manifest nodes, it is not a problem because the `tags` property is not implemented in the other `SelectorTarget` classes.

### Fix

In our code, creating nodes of type `UnitTestDefinitions` for unit tests, instead of `_DictShim`, fixes the problem.

## How I Tested These Changes

Tested locally + additional test in BK

### CODE REVIEW

**Comment by maximearmstrong** [comment]
* **#23232** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23232?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @maximearmstrong and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/libraries/dagster-dbt/dagster_dbt_tests/core/test_asset_decorator.py [hunk 1]
```diff
@@ -1060,6 +1060,27 @@ def my_dbt_assets(context: AssetExecutionContext, dbt: DbtCliResource):
     assert result.success
 
 
+@pytest.mark.skipif(
+    version.parse(dbt_version) < version.parse("1.8.0"),
+    reason="dbt unit test support is only available in `dbt-core>=1.8.0`",
+)
+def test_dbt_with_unit_tests_with_tags_selector(
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt_tests/core/test_asset_decorator.py:1067]: can we just parametrize the test above by `select`?

Then we have two options: `select: ["fqn:*", "tag:test"]`

File: python_modules/libraries/dagster-dbt/dagster_dbt/utils.py [hunk 2]
```diff
@@ -40,6 +41,23 @@ def __getattr__(self, item):
...
+                    # unit test nodes must be of type UnitTestDefinition
+                    unique_id: UnitTestDefinition(**info)
+                    for unique_id, info in manifest_json["unit_tests"].items()
+                },
+            }
+            if manifest_json.get("unit_tests")
+            else {}
+        )
+    else:
+        unit_test = {}
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/utils.py:59]: nit: remove `else`, use `version.parse`, rename to `unit_tests`

```suggestion
    unit_tests = {}
    if version.parse(dbt_version) >= version.parse("1.8.0"):
        from dbt.contracts.graph.nodes import UnitTestDefinition

        unit_tests = (
            {
                "unit_tests": {
                    # unit test nodes must be of type UnitTestDefinition
                    unique_id: UnitTestDefinition(**unit_test_resource_props)
                    for unique_id, unit_test_resource_props in manifest_json["unit_tests"].items()
                },
            }
            if manifest_json.get("unit_tests")
            else {}
        )
```

File: python_modules/libraries/dagster-dbt/dagster_dbt_tests/dbt_projects/test_dagster_dbt_unit_tests/models/staging/stg_payments.sql [hunk 3]
```diff
@@ -1,3 +1,9 @@
+{{
+    config(
+        tags=["test"],
+    )
+}}
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt_tests/dbt_projects/test_dagster_dbt_unit_tests/models/staging/stg_payments.sql:5]: maybe apply the tags to all models in `dbt_project.yml`? I noticed we didn't apply the tag to `customers.sql`, which is the model that has the unit test. I would want to see the interaction with a selection on a model with a unit test

**Review by rexledesma — APPROVED** [review]
see comments!

File: python_modules/libraries/dagster-dbt/dagster_dbt_tests/dbt_projects/test_dagster_dbt_unit_tests/models/staging/stg_payments.sql (hunk 3 above)
Comment (maximearmstrong) [reply python_modules/libraries/dagster-dbt/dagster_dbt_tests/dbt_projects/test_dagster_dbt_unit_tests/models/staging/stg_payments.sql]: Done in [1d3daf4](https://github.com/dagster-io/dagster/pull/23232/commits/1d3daf41f8a1b2171a10a7088cea81101dc346f4)

File: python_modules/libraries/dagster-dbt/dagster_dbt_tests/core/test_asset_decorator.py (hunk 1 above)
Comment (maximearmstrong) [reply python_modules/libraries/dagster-dbt/dagster_dbt_tests/core/test_asset_decorator.py]: Done in [5d3fee4](https://github.com/dagster-io/dagster/pull/23232/commits/5d3fee41fb2333d37016a8b5f7cd565d8c875774)

File: python_modules/libraries/dagster-dbt/dagster_dbt/utils.py (hunk 2 above)
Comment (maximearmstrong) [reply python_modules/libraries/dagster-dbt/dagster_dbt/utils.py]: Done in [88e20dd](https://github.com/dagster-io/dagster/pull/23232/commits/88e20ddec7762b6e75ad918ef9ec7b02cb4c2f09)

**Pull request merged**